kratos/mpi/includes/mpi_communicator.h

```diff
@@ -74,673 +67,978 @@ namespace Kratos
+{
 
-                int send_tag = i_color;
-                int receive_tag = i_color;
+const Variable<TValue>& mrVariable;
 
-                MPI_Sendrecv(send_buffer, send_buffer_size, MPI_INT, destination, send_tag, receive_buffer, receive_buffer_size, MPI_INT, destination, receive_tag,
-                             MPI_COMM_WORLD, &status);
+public:
 
-                // Updating nodes
-                position = 0;
-                for (ModelPart::NodeIterator i_node = GhostMesh(i_color).NodesBegin();
-                        i_node != GhostMesh(i_color).NodesEnd(); i_node++)
-                    for (ModelPart::NodeType::DofsContainerType::iterator i_dof = i_node->GetDofs().begin(); i_dof != i_node->GetDofs().end(); i_dof++)
-                        i_dof->SetEquationId(receive_buffer[position++]);
+using ValueType = TValue;
+using SendType = typename SendTools<TValue>::SendType;
+using ContainerType = Communicator::MeshType::ElementsContainerType;
+using IteratorType = Communicator::MeshType::ElementsContainerType::iterator;
+using ConstIteratorType = Communicator::MeshType::ElementsContainerType::const_iterator;
 
-                if (position > receive_buffer_size)
-                    std::cout << rank << " Error in estimating receive buffer size...." << std::endl;
+ElementalDataAccess(const Variable<TValue>& mrVariable):
+    mrVariable(mrVariable)
+{}
 
-                delete [] send_buffer;
-                delete [] receive_buffer;
-            }
+ContainerType& GetContainer(Communicator::MeshType& rMesh)
+{
+    return rMesh.Elements();
+}
+
+const ContainerType& GetContainer(const Communicator::MeshType& rMesh)
+{
+    return rMesh.Elements();
+}
+
+TValue& GetValue(IteratorType& iter)
+{
+    return iter->GetValue(mrVariable);
+}
+
+const TValue& GetValue(const ConstIteratorType& iter)
+{
+    return iter->GetValue(mrVariable);
+}
+
+};
+
+class ElementalFlagsAccess
+{
+public:
+
+const Kratos::Flags& mrMask;
+
+using ValueType = Kratos::Flags;
+using SendType = typename SendTools<ValueType>::SendType;
+using ContainerType = Communicator::MeshType::ElementsContainerType;
+using IteratorType = Communicator::MeshType::ElementsContainerType::iterator;
+using ConstIteratorType = Communicator::MeshType::ElementsContainerType::const_iterator;
+
+ElementalFlagsAccess(const Kratos::Flags& rMask):
+    mrMask(rMask)
+{}
+
+ContainerType& GetContainer(Communicator::MeshType& rMesh)
+{
+    return rMesh.Elements();
+}
+
+const ContainerType& GetContainer(const Communicator::MeshType& rMesh)
+{
+    return rMesh.Elements();
+}
+
+Kratos::Flags& GetValue(IteratorType& iter)
+{
+    return *iter;
+}
+
+const Kratos::Flags& GetValue(const ConstIteratorType& iter)
+{
+    return *iter;
+}
+
+};
+
+}
+
+/// MPICommunicator manages the transfer of ModelPart data in MPI distributed memory environment.
+class MPICommunicator : public Communicator
+{
+
+enum class DistributedType {
+    Local,
+    Ghost
+};
```

Comment:
and Interface?

Reply:
I did not add it because there is no function using the entire interface currently, so I had no test for it.